src/stack/Stack.h

```diff
@@ -0,0 +1,264 @@
+
+    CHIP_ERROR Shutdown() { return mSystemLayer.Shutdown(); }
+
+    System::Layer & Get() { return mSystemLayer; }
+
+private:
+    System::Layer mSystemLayer;
+};
+#endif // CONFIG_DEVICE_LAYER
+
+/* ========================== InetLayer Configuration ========================== */
+struct InetLayerConfiguration
+{
+};
+template <typename T>
+struct IsInetLayerConfiguration : std::is_base_of<InetLayerConfiguration, T>
+{
+};
+
+#if CONFIG_DEVICE_LAYER
+class DefaultInetLayer : InetLayerConfiguration
+{
+public:
+    CHIP_ERROR Init(System::Layer & aSystemLayer) { return CHIP_NO_ERROR; }
+    CHIP_ERROR Shutdown() { return CHIP_NO_ERROR; }
+
+    Inet::InetLayer & Get() { return DeviceLayer::InetLayer; }
+};
+#else
+class DefaultInetLayer : InetLayerConfiguration
+{
+public:
+    CHIP_ERROR Init(System::Layer & aSystemLayer) { return mInetLayer.Init(aSystemLayer, nullptr); }
+
+    CHIP_ERROR Shutdown() { return mInetLayer.Shutdown(); }
+
+    Inet::InetLayer & Get() { return mInetLayer; }
+
+private:
+    Inet::InetLayer mInetLayer;
+};
+#endif // CONFIG_DEVICE_LAYER
+
+/* ========================== Storage Configuration ========================== */
+struct StorageConfiguration
+{
+};
+template <typename T>
+struct IsStorageConfiguration : std::is_base_of<StorageConfiguration, T>
+{
+};
+
+class DefaultStorage : StorageConfiguration
+{
+public:
+    CHIP_ERROR Init() { return CHIP_NO_ERROR; }
+    CHIP_ERROR Shutdown() { return CHIP_NO_ERROR; }
+
+    PersistentStorageDelegate * Get() { return nullptr; }
+};
+
+/* ========================== BleLayer Configuration ========================== */
+struct BleLayerConfiguration
+{
+};
+template <typename T>
+struct IsBleLayerConfiguration : std::is_base_of<BleLayerConfiguration, T>
+{
+};
+
+class DefaultBleLayer : BleLayerConfiguration
+{
+public:
+    template <typename T>
+    CHIP_ERROR Init(T * stack)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    Ble::BleLayer * Get() { return nullptr; }
+};
+
+/* ========================== Transport Configuration ========================== */
+struct TransportConfiguration
+{
+};
+template <typename T>
+struct IsTransportConfiguration : std::is_base_of<TransportConfiguration, T>
+{
+};
+
+class DefaultTransport : TransportConfiguration
+{
+public:
+    using transport = TransportMgr<
+#if INET_CONFIG_ENABLE_IPV4
+        Transport::UDP, /* IPv4 */
+#endif
+        Transport::UDP>;
+
+    CHIP_ERROR Init(Inet::InetLayer & inetLayer, Ble::BleLayer * bleLayer)
+    {
+        return mTransportManager.Init(
+#if INET_CONFIG_ENABLE_IPV4
+            Transport::UdpListenParameters(&inetLayer).SetAddressType(Inet::kIPAddressType_IPv4).SetListenPort(mPort),
+#endif
+            Transport::UdpListenParameters(&inetLayer).SetAddressType(Inet::kIPAddressType_IPv6).SetListenPort(mPort));
+    }
+
+    TransportMgrBase & Get() { return mTransportManager; }
+    void SetListenPort(uint16_t port) { mPort = port; }
```

Comment:
Since this is used during init, why are we storing this as a member? Could we pass this as part of the init arguments instead?

Reply:
As a user, it sounds like I may want some 'StackInitParameters' to define how the stack is brought up. The listening port seems to be part of that.

I would rather have `gStack.Init(Args().SetListenPort(1234))` or maybe:

```cpp
StackParameters params;
params.SetListenPort(1234);

gStack.init(params)

```

than:
```cpp
gStack.GetTransportConfig().SetListenPort(1234);
gStack.Init();
```

because the latter relies on ordering and how arguments are passed around is not immediately clear.